### PR TITLE
use package-scheme imports for generated web entrypoint

### LINKED ISSUE #43846: VM and dart2js have different behavior in the presence of combined package/relative imports
## Steps to Reproduce

1. Imported a class in my Flutter application in different locations using different paths:
```
/// file A
import service/google_authentication.dart
/// file B
import package:app_flutter/service/google_authentication.dart
```
2. `flutter build web` 
  - fails as the Google Authentication already exists
  - `flutter build aot` builds successfully.

**Target Platform:** Web
**Target OS version/browser:** Chrome
**Devices:** N/A

Discovered in https://github.com/flutter/cocoon/pull/499 

To quote @jonahwilliams: It seems like the frontend_server applies some sort of canonicalization, while dart2js does not. The fix will likely be to teach the frontend_server to generate dart2js compatible kernel and use that instead.

## Logs

```
$ flutter build web

#0      Dart2JSTarget.build (package:flutter_tools/src/build_system/targets/web.dart:154:7)
<asynchronous suspension>
#1      _BuildInstance._invokeInternal (package:flutter_tools/src/build_system/build_system.dart:526:25)
<asynchronous suspension>
#2      _BuildInstance.invokeTarget.<anonymous closure> (package:flutter_tools/src/build_system/build_system.dart:481:35)
#3      new Future.sync (dart:async/future.dart:222:31)
#4      AsyncMemoizer.runOnce (package:async/src/async_memoizer.dart:43:45)
#5      _BuildInstance.invokeTarget (package:flutter_tools/src/build_system/build_system.dart:481:21)
<asynchronous suspension>
<asynchronous suspension>
#6      BuildSystem.build (package:flutter_tools/src/build_system/build_system.dart:419:36)
#7      _AsyncAwaitCompleter.start (dart:async-patch/async_patch.dart:43:6)
#8      BuildSystem.build (package:flutter_tools/src/build_system/build_system.dart:400:28)
#9      buildWeb (package:flutter_tools/src/web/compile.dart:34:50)
#10     _asyncThenWrapperHelper.<anonymous closure> (dart:async-patch/async_patch.dart:71:64)
#11     _rootRunUnary (dart:async/zone.dart:1132:38)
#12     _CustomZone.runUnary (dart:async/zone.dart:1029:19)
#13     _FutureListener.handleValue (dart:async/future_impl.dart:137:18)
#14     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:678:45)
#15     Future._propagateToListeners (dart:async/future_impl.dart:707:32)
#16     Future._completeWithValue (dart:async/future_impl.dart:522:5)
#17     _AsyncAwaitCompleter.complete (dart:async-patch/async_patch.dart:30:15)
#18     _completeOnAsyncReturn (dart:async-patch/async_patch.dart:288:13)
#19     injectPlugins (package:flutter_tools/src/plugins.dart)
#20     _asyncThenWrapperHelper.<anonymous closure> (dart:async-patch/async_patch.dart:71:64)
#21     _rootRunUnary (dart:async/zone.dart:1132:38)
#22     _CustomZone.runUnary (dart:async/zone.dart:1029:19)
#23     _FutureListener.handleValue (dart:async/future_impl.dart:137:18)
#24     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:678:45)
#25     Future._propagateToListeners (dart:async/future_impl.dart:707:32)
#26     Future._completeWithValue (dart:async/future_impl.dart:522:5)
#27     Future._asyncComplete.<anonymous closure> (dart:async/future_impl.dart:552:7)
#28     _rootRun (dart:async/zone.dart:1124:13)
#29     _CustomZone.run (dart:async/zone.dart:1021:19)
#30     _CustomZone.runGuarded (dart:async/zone.dart:923:7)
#31     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:963:23)
#32     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
#33     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
#34     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:116:13)
#35     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:173:5)


Exception: lib/build_dashboard.dart:81:56:                              
Error: A value of type 'GoogleSignInService/*1*/' can't be assigned to a variable of type 'GoogleSignInService/*2*/'.
 - 'GoogleSignInService/*1*/' is from 'lib/service/google_authentication.dart'.
 - 'GoogleSignInService/*2*/' is from 'package:app_flutter/service/google_authentication.dart' ('lib/service/google_authentication.dart').
    final GoogleSignInService authService = buildState.authService;
                                                       ^
Error: Compilation failed.
```

```
chillers@chillers:~/cocoon/app_flutter$ flutter doctor -v
[✓] Flutter (Channel master, v1.10.15-pre.320, on Linux, locale en_US.UTF-8)
    • Flutter version 1.10.15-pre.320 at /usr/local/google/home/chillers/flutter
    • Framework revision 3352b91f2c (22 hours ago), 2019-10-29 14:31:27 -0700
    • Engine revision 419f5d594a
    • Dart version 2.6.0 (build 2.6.0-dev.8.2 e1fce75301)

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /usr/local/google/home/chillers/Android/Sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-stable, build-tools 29.0.2
    • Java binary at: /opt/android-studio-with-blaze-3.4/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b16-5323222)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[!] Android Studio (version 3.4)
    • Android Studio at /opt/android-studio-with-blaze-3.4
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b16-5323222)

[✓] IntelliJ IDEA Ultimate Edition (version 2019.1)
    • IntelliJ at /opt/intellij-ue-2019.1
    • Flutter plugin version 39.0.2
    • Dart plugin version 191.8423

[✓] VS Code (version 1.39.2)
    • VS Code at /usr/share/code
    • Flutter extension version 3.5.1

[✓] Connected device (3 available)
    • Android SDK built for x86 • emulator-5554 • android-x86    • Android 10 (API 29) (emulator)
    • Chrome                    • chrome        • web-javascript • Google Chrome 78.0.3904.70
    • Web Server                • web-server    • web-javascript • Flutter Tools

! Doctor found issues in 1 category.
```

### LINKED ISSUE #44111: [web] what are differences between flutter run and simple http server of web build
I'm currently developing a flutter UI for a touch panel, and the app works great if I use "flutter run -d chrome".  If I run it though from a simple http server I get what look like minification lookup failures?  Here is how I'm starting a simple server:

`python3 -m http.server 8000
`
from the 'build/web' path.  Here is how I'm doing the web build:

`flutter build web`

when I run it this way, I get a failure like this:

image

So my first question is, why would running this with "flutter run -d chrome" work better?  What's different about the build that way?

My next question, is there some better way to get failure logs from a web UI?  From the console I see this:

image

Here is my flutter info:

```
JustinMac:audacy_on_tap jeltoft$ flutter doctor -v
[✓] Flutter (Channel unknown, v1.10.15-pre.352, on Mac OS X 10.14.6 18G103, locale en-US)
    • Flutter version 1.10.15-pre.352 at /Users/jeltoft/flutter/flutter
    • Framework revision 01d55074af (4 days ago), 2019-10-31 10:10:25 -0500
    • Engine revision b16eab6aff
    • Dart version 2.6.0 (build 2.6.0-dev.8.2 d50c158d42)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/jeltoft/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 10.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.2.1, Build version 10E1001
    • CocoaPods version 1.7.5

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 40.2.2
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] Connected device (2 available)
    • Chrome     • chrome     • web-javascript • Google Chrome 78.0.3904.70
    • Web Server • web-server • web-javascript • Flutter Tools

• No issues found!
```

### LINKED ISSUE #44600: [web] 'flutter build web' and 'flutter run' have different behavior related to type resolution
This issue is specific to __flutter web__.

# Summary

There are two ways to reference a file when importing from a local package: `import "abc.dart"` and `import "package:my_package/abc.dart"`. As I found out, a class declared in `abc.dart` and referenced from different files using the two different ways results in two types that are not equivalent, where the `is` operator returns `false`. That happens only if the app is built using `flutter build web`. When running in debug mode using `flutter run -d chrome` the app behaves correctly.

# To reproduce

1. Make sure you're on master branch of Flutter.
1. Create a new Flutter project.
2. Copy-paste `main.dart`, `model.dart`, `util.dart` into your project lib directory.
3. Replace `my_package` with the name of your package in `main.dart` and `util.dart`.
4. Run the command: `flutter run -d chrome`. A Chrome window will open with `true` printed in the center.
5. Run the command `flutter build web`. 
6. Run the command `http-server build/web -o`. A Chrome window will open with `false` printed in the center.

# Sources

## main.dart
```dart
import 'package:flutter/material.dart';
import 'package:my_package/util.dart';
import 'model.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Bug Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  @override
  Widget build(BuildContext context) {

    final m1 = createModel();

    final isModel = m1 is MyModel;

    return Scaffold(
      appBar: AppBar(
        title: Text('Bug Demo'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              "If the displayed value is 'false' then it's a bug",
            ),
            Text(
              '$isModel',
              style: Theme.of(context).textTheme.display1,
            ),
          ],
        ),
      ),
    );
  }
}
```

## model.dart
```dart
class MyModel {
}
```

## util.dart
```dart
import 'package:my_package/model.dart';

Object createModel() {
  return new MyModel();
}
```

# Flutter doctor
```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel master, v1.10.15-pre.248, on Mac OS X 10.14.6 18G95, locale en-US)
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[✓] Xcode - develop for iOS and macOS (Xcode 11.0)
[✓] Chrome - develop for the web
[✓] Android Studio (version 3.5)
[✓] VS Code (version 1.39.2)
[✓] Connected device (3 available)

• No issues found!
```

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Uri fragments are resolving to file Uris, and the mix of the file uris and package uris is causing type errors since the same types have different identities based on how they are imported. Import the main package with a package scheme to fix the issue.

Verified with posted examples that this was sufficient to fix the issue.

Fixes https://github.com/flutter/flutter/issues/44111
Fixes https://github.com/flutter/flutter/issues/44600
Fixes https://github.com/flutter/flutter/issues/43846

### CODE REVIEW

**Comment by jonahwilliams** [comment]
FYI @sigmundch

File: packages/flutter_tools/lib/src/build_system/targets/web.dart [hunk 1]
```diff
@@ -51,21 +52,29 @@ class WebEntrypointTarget extends Target {
     final String targetFile = environment.defines[kTargetFile];
     final bool shouldInitializePlatform = environment.defines[kInitializePlatform] == 'true';
     final bool hasPlugins = environment.defines[kHasWebPlugins] == 'true';
-    final String import = fs.file(fs.path.absolute(targetFile)).uri.toString();
+    final String importPath = fs.path.absolute(targetFile);
+    final PackageUriMapper packageUriMapper = PackageUriMapper(
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/build_system/targets/web.dart]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#if-you-have-a-newline-after-some-opening-punctuation-match-it-on-the-closing-punctuation

File: packages/flutter_tools/lib/src/build_system/targets/web.dart [hunk 2]
```diff
@@ -51,21 +52,29 @@ class WebEntrypointTarget extends Target {
...
+      final Uri generatedImport = packageUriMapper.map(generatedPath);
       contents = '''
 import 'dart:ui' as ui;
 
 import 'package:flutter_web_plugins/flutter_web_plugins.dart';
 
-import '$generatedPath';
-import "$import" as entrypoint;
+import '$generatedImport';
+import "$mainImport" as entrypoint;
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/build_system/targets/web.dart]: Can this import use single quotes?

File: packages/flutter_tools/lib/src/build_system/targets/web.dart [hunk 3]
```diff
@@ -79,7 +88,7 @@ Future<void> main() async {
       contents = '''
 import 'dart:ui' as ui;
 
-import "$import" as entrypoint;
+import "$mainImport" as entrypoint;
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/build_system/targets/web.dart]: ditto

**Review by zanderso — COMMENTED** [review]
Just nits

File: packages/flutter_tools/lib/src/build_system/targets/web.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_system/targets/web.dart]: Fixed

File: packages/flutter_tools/lib/src/build_system/targets/web.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_system/targets/web.dart]: Done

File: packages/flutter_tools/lib/src/build_system/targets/web.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_system/targets/web.dart]: Done

**Review by yjbanov — APPROVED** [review]

File: packages/flutter_tools/lib/src/build_system/targets/web.dart [hunk 4]
```diff
@@ -51,21 +52,32 @@ class WebEntrypointTarget extends Target {
...
-    final String import = fs.file(fs.path.absolute(targetFile)).uri.toString();
+    final String importPath = fs.path.absolute(targetFile);
+    final PackageUriMapper packageUriMapper = PackageUriMapper(
+      importPath,
+      PackageMap.globalPackagesPath,
+      null,
+      null,
+    );
+    final Uri mainImport = packageUriMapper.map(importPath);
+    if (mainImport == null) {
```
Comment (sigmundch) [inline packages/flutter_tools/lib/src/build_system/targets/web.dart]: technically, this is not a problem: we only have a problem when it's possible to import the same file via `package:*` imports.

File: packages/flutter_tools/lib/src/build_system/targets/web.dart [hunk 5]
```diff
@@ -51,21 +52,32 @@ class WebEntrypointTarget extends Target {
     final String targetFile = environment.defines[kTargetFile];
     final bool shouldInitializePlatform = environment.defines[kInitializePlatform] == 'true';
     final bool hasPlugins = environment.defines[kHasWebPlugins] == 'true';
-    final String import = fs.file(fs.path.absolute(targetFile)).uri.toString();
+    final String importPath = fs.path.absolute(targetFile);
+    final PackageUriMapper packageUriMapper = PackageUriMapper(
```
Comment (sigmundch) [inline packages/flutter_tools/lib/src/build_system/targets/web.dart]: would it help to add a one-line comment explaining why we need to use the package import?

**Review by sigmundch — COMMENTED** [review]
Thanks for the quick fix!

It would be nice to also create a regression test that would break if we ever switch accidentally to stop using the package-import. Is that possible to do on your testing infrastructure?

**Comment by jonahwilliams** [comment]
I agree wrt the regression test, but we don't yet have a way to interact with a flutter web app that is compiled for release mode, but hopefully soon. Filled https://github.com/flutter/flutter/issues/45995

**Pull request merged**